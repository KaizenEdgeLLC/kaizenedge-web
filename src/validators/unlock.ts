import type { FitnessOnboarding, UnlockEvaluation, UnlockFlag, PremiumTheme } from "../../types/types";

export function computeUnlocks(input: FitnessOnboarding): UnlockEvaluation {
  const flags = new Set<UnlockFlag>();
  const reasons: string[] = [];

  const tf = input.timeFrequency;
  if (tf.daysPerWeek >= 4) { flags.add("freq_4plus"); reasons.push("Trains ≥4 days/week"); }
  if (tf.intensityPreference === "high") { flags.add("intensity_high"); reasons.push("High intensity preference"); }

  const env = input.environment;
  if (env?.stylePreferences?.some(s => s === "martial_arts" || s === "calisthenics")) {
    flags.add("style_martial_or_calisthenics");
    reasons.push("Style preference includes martial arts or calisthenics");
  }

  const goals = input.goals;
  if ((goals.targetTimelineDays ?? 0) >= 90) {
    flags.add("timeline_90plus");
    reasons.push("Target timeline ≥ 90 days");
  }

  const nut = input.nutrition;
  if (nut?.culturalFlavors?.some(f => f === "japan" || f === "korea")) {
    flags.add("flavor_japan_or_korea");
    reasons.push("Cultural flavors include Japan/Korea");
  }
  if (nut?.snackStyle === "gamer_snacks") {
    flags.add("snack_gamer");
    reasons.push("Snack style suggests gamer snacks");
  }
  if (nut?.timingStyle === "night_owl") {
    flags.add("timing_night_owl");
    reasons.push("Night-owl timing");
  }

  const beh = input.behavior;
  if (beh?.gamificationOptIn) {
    flags.add("gamification_on");
    reasons.push("Gamification opt-in enabled");
  }
  const tooEasyCount = beh?.sessionFeedbackHistory?.filter(s => s.perceivedDifficulty === "too_easy").length ?? 0;
  if (tooEasyCount >= 2) {
    flags.add("too_easy_trend");
    reasons.push("Multiple 'too easy' feedback entries");
  }

  if (input.integrations?.devices && input.integrations.devices.some(d => d !== "none")) {
    flags.add("tech_biofeedback_interest");
    reasons.push("Health device integrations connected");
  }

  const flagCount = flags.size;
  const ascensionCandidate = flagCount >= 3;
  const otherCandidates: PremiumTheme[] = [];

  if (input.behavior?.motivationArchetype === "explorer" || env?.trainingLocation === "outdoor") {
    otherCandidates.push("adventure");
  }
  if ((env?.stylePreferences ?? []).includes("yoga") || input.nutrition?.dietaryPattern === "vegetarian" || input.timeFrequency.impactLevel === "low") {
    otherCandidates.push("zen");
  }
  if ((input.integrations?.devices ?? []).some(d => d === "apple_health" || d === "garmin") && (input.integrations?.metricsIngest ?? []).length) {
    otherCandidates.push("cybernetic");
  }

  return {
    flags: Array.from(flags),
    flagCount,
    ascensionCandidate,
    otherCandidates: otherCandidates.filter((v, i, a) => a.indexOf(v) === i && v !== "ascension") as UnlockEvaluation["otherCandidates"],
    unlockReason: reasons
  };
}
