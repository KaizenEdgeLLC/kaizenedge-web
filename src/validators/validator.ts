import Ajv, { ErrorObject } from "ajv";
import addFormats from "ajv-formats";
import schema from "../../schemas/fitness-onboarding.v1.json";
import type { FitnessOnboarding } from "../../types/types";

const ajv = new Ajv({ strict: true, allErrors: true });
addFormats(ajv);

export const validateOnboarding = ajv.compile<FitnessOnboarding>(schema);

export function assertOnboarding(input: unknown): FitnessOnboarding {
  const ok = validateOnboarding(input);
  if (!ok) {
    const errs = (validateOnboarding.errors ?? []).map(formatAjvError).join("\n");
    throw new Error(`Onboarding payload failed validation:\n${errs}`);
  }
  return input as FitnessOnboarding;
}

function formatAjvError(e: ErrorObject): string {
  const path = e.instancePath && e.instancePath.length ? e.instancePath : "(root)";
  const msg = e.message ?? "";
  const params = e.params ? ` ${JSON.stringify(e.params)}` : "";
  return `â€¢ ${path} ${msg}${params}`.trim();
}
