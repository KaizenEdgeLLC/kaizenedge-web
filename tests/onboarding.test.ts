import { assertOnboarding } from "../src/validators/validator";
import { computeUnlocks } from "../src/validators/unlock";
import type { FitnessOnboarding } from "../types/types";

const payload: FitnessOnboarding = {
  meta: { version: "1.0.0", timestamp: new Date().toISOString(), source: "ios" },
  userProfile: {
    age: 35, heightCm: 178, weightKg: 86,
    biologicalSex: "unspecified",
    restingHrBpm: 58,
    conditions: [], injuries: ["knee"],
    medicalClearance: "self-cleared"
  },
  goals: { primary: "strength", secondary: ["mobility"], targetTimelineDays: 120 },
  environment: {
    trainingLocation: "hybrid",
    equipment: ["dumbbells", "yoga_mat"],
    stylePreferences: ["calisthenics", "hiit", "yoga"],
    culturalTags: ["brazil_capoeira", "india_yoga"]
  },
  timeFrequency: {
    minutesPerSession: 45, daysPerWeek: 5,
    restStrategy: "auto", intensityPreference: "high", impactLevel: "medium"
  },
  nutrition: {
    dietaryPattern: "omnivore",
    culturalFlavors: ["korea", "japan"],
    timingStyle: "night_owl",
    snackStyle: "gamer_snacks",
    proteinTargetGPerDay: 150,
    calorieTargetKcalPerDay: 2600
  },
  integrations: { devices: ["apple_health"], metricsIngest: ["heart_rate", "sleep", "steps"] },
  behavior: {
    motivationArchetype: "competitor",
    gamificationOptIn: true,
    notificationStyle: "system_update",
    streakTargetDays: 30,
    sessionFeedbackHistory: [
      { date: "2025-08-20", perceivedDifficulty: "too_easy" },
      { date: "2025-08-22", perceivedDifficulty: "just_right" },
      { date: "2025-08-24", perceivedDifficulty: "too_easy" }
    ]
  }
};

describe("KaizenEdge Fitness Onboarding â€” validation & unlocks", () => {
  test("validates and computes unlocks deterministically", () => {
    const data = assertOnboarding(payload);
    const unlocks = computeUnlocks(data);

    expect(unlocks.flagCount).toBeGreaterThanOrEqual(3);
    expect(unlocks.flags).toEqual(
      expect.arrayContaining([
        "freq_4plus",
        "intensity_high",
        "style_martial_or_calisthenics",
        "flavor_japan_or_korea",
        "timing_night_owl",
        "snack_gamer",
        "gamification_on",
        "too_easy_trend",
        "tech_biofeedback_interest"
      ])
    );
    expect(unlocks.ascensionCandidate).toBe(true);
    expect(unlocks.unlockReason && unlocks.unlockReason.length).toBeGreaterThan(0);

    expect(unlocks).toMatchSnapshot();
  });

  test("invalid payload fails validation", () => {
    const bad: any = { meta: { version: "1.0.0" } };
    expect(() => assertOnboarding(bad)).toThrow(/Onboarding payload failed validation/);
  });
});
