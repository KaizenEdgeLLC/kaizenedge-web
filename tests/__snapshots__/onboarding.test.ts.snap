// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`KaizenEdge Fitness Onboarding — validation & unlocks validates and computes unlocks deterministically 1`] = `
{
  "ascensionCandidate": true,
  "flagCount": 10,
  "flags": [
    "freq_4plus",
    "intensity_high",
    "style_martial_or_calisthenics",
    "timeline_90plus",
    "flavor_japan_or_korea",
    "snack_gamer",
    "timing_night_owl",
    "gamification_on",
    "too_easy_trend",
    "tech_biofeedback_interest",
  ],
  "otherCandidates": [
    "zen",
    "cybernetic",
  ],
  "unlockReason": [
    "Trains ≥4 days/week",
    "High intensity preference",
    "Style preference includes martial arts or calisthenics",
    "Target timeline ≥ 90 days",
    "Cultural flavors include Japan/Korea",
    "Snack style suggests gamer snacks",
    "Night-owl timing",
    "Gamification opt-in enabled",
    "Multiple 'too easy' feedback entries",
    "Health device integrations connected",
  ],
}
`;
